src/arm-codegen.c

```diff
@@ -294,7 +318,25 @@ void emit_ph2_ir(ph2_ir_t *ph2_ir)
         return;
     case OP_call:
         func = find_func(ph2_ir->func_name);
-        emit(__bl(__AL, func->bbs->elf_offset - elf_code->size));
+        if (func->bbs)
+            ofs = func->bbs->elf_offset - elf_code->size;
+        else if (dynlink) {
+            ofs = (dynamic_sections.elf_plt_start + func->plt_offset) -
+                  (elf_code_start + elf_code->size + 4);
+            is_external_call = true;
+        } else {
+            printf("The '%s' function is not implemented\n", ph2_ir->func_name);
+            abort();
+        }
+
+        /* If the callee is external, save __r12 at [sp + 16] and
+         * restore it after the function returns.
+         */
+        if (is_external_call)
+            emit(__sw(__AL, __r12, __sp, 16));
+        emit(__bl(__AL, ofs));
+        if (is_external_call)
+            emit(__lw(__AL, __r12, __sp, 16));
```

Comment:
Why save r12?
- r12 (IP) is a scratch register in AAPCS, external functions can freely modify it
- shecc uses r12 to store the global stack pointer
- Therefore, every external call must save/restore it

Consolidate comments.